src/main/typescript/wcardinal/ui/d-tree.ts

```diff
@@ -3,19 +3,400 @@
+		protected _itemIndexMappedEnd!: number;
+		private _itemY!: number;
+		private _itemHeight!: number;
+		private _removeItem!: DTreeItemRawData | null;
+		private _addItemOptions!: DTreeAddedItemOptions | null;
+
+		protected init(options ?: OPTIONS) {
+			super.init(options);
+
+			this._itemOptions = new WeakMap();
+			this._selection = new DTreeSelection();
+			this._itemOptionsShowable = [];
+			this._itemIndexMappedStart = 0;
+			this._itemIndexMappedEnd = 0;
+			this._itemY = 0;
+
+			const dTreeItemTheme: DThemeTreeItem = DThemes.getInstance().get("DTreeItem");
+			this._itemHeight = Number(dTreeItemTheme.getHeight());
+
+			this._value = options && options.value ? options.value : [];
+			this.updateData(null, this, 0);
+
+			this._content.on("move", (): void => {
+				this.update();
+			});
+
+			this._content.on("resize", (): void => {
+				this._content.removeChildren();
+				this.update();
+			});
+
+			this.onSelectionChangeListener();
+
+			this.update();
+		}
+
+		private onSelectionChangeListener() {
+			this._content.removeListener("selection-change");
+			// update active state of all shown item when a child item is clicked
+			this._content.on("selection-change", () => {
+				const items = this._content.children as DTreeItem[];
+				items.forEach((item) => {
+					item.updateActiveState(this._selection.get().includes(item.getRawData()));
+				});
+			});
+		}
+
+		protected update() {
+			const content = this._content;
+			const items = content.children as DTreeItem[] ;
+			// calculate content height
+			content.height = this._itemOptionsShowable.length * this._itemHeight;
+			const contentY = content.position.y;
+			const height = this.height;
+			const itemIndexMappedStart = (0 - contentY) / this._itemHeight - 2 < 0 ?
+				0 :
+				Math.floor((0 - contentY) / this._itemHeight) - 2;
+
+			const itemIndexMappedEnd = (height - contentY) / this._itemHeight + 2 < this._itemOptionsShowable.length - 1 ?
+				Math.floor((height - contentY) / this._itemHeight) + 2 :
+				this._itemOptionsShowable.length;
+
+			// get items options are shown in Dpane content frame
+			const itemOptionsShown = this._itemOptionsShowable.slice(itemIndexMappedStart, itemIndexMappedEnd);
+
+			if (items.length < itemOptionsShown.length) {
+				for (let i = items.length; i < itemOptionsShown.length; i++) {
+					const itemOptions = itemOptionsShown[i];
+					const treeItem = new DTreeItem(itemOptions);
+					content.addChild(treeItem);
+					// listen select item event
+					treeItem.on("select", (): void => {
+						this._selection.clear();
+						this._selection.add(treeItem.getRawData());
+					});
+					// listen toggle item event
+					treeItem.on("toggle", (): void => {
+						if(treeItem.isParent()) {
+							this.toggle(treeItem.getRawData());
+						}
+					});
+				}
+			} else if (items.length > itemOptionsShown.length) {
+				for (let i = itemOptionsShown.length; i < items.length; i++) {
+					content.removeChild(items[i]);
```

Comment:
`removeChild` is a relatively heavy operation. If possible, please avoid using it.  Please consider using `items[ i ].visible = false` for example.

Reply:
Thank you for your suggestions. I will change this code.